dapr_agents/agents/durableagent/agent.py

```diff
@@ -118,148 +139,129 @@ def tool_calling_workflow(self, ctx: DaprWorkflowContext, message: TriggerAction
+        )
+
+        source = entry_info.get("source")
+        source_workflow_instance_id = entry_info.get("source_workflow_instance_id")
 
-        # Step 3: Generate Response
+        # Step 3: Generate Response via LLM
         response = yield ctx.call_activity(
-            self.generate_response, input={"instance_id": instance_id, "task": task}
+            self.generate_response, input={"task": task, "instance_id": instance_id}
         )
+
+        # Step 4: Extract Response Message from LLM Response
         response_message = yield ctx.call_activity(
             self.get_response_message, input={"response": response}
         )
 
-        # Step 4: Extract Finish Reason
+        # Step 5: Extract Finish Reason from LLM Response
         finish_reason = yield ctx.call_activity(
             self.get_finish_reason, input={"response": response}
         )
 
-        # Step 5: Choose execution path based on LLM response
-        if finish_reason == "tool_calls":
-            if not ctx.is_replaying:
-                logger.info(
-                    "Tool calls detected in LLM response, extracting and preparing for execution.."
-                )
+        # Step 6:Add the assistant's response message to the chat history
+        yield ctx.call_activity(
+            self.append_assistant_message,
+            input={"instance_id": instance_id, "message": response_message},
+        )
 
+        # Step 7: Handle tool calls response
+        if finish_reason == FinishReason.TOOL_CALLS:
+            if not ctx.is_replaying:
+                logger.info("Tool calls detected in LLM response.")
             # Retrieve the list of tool calls extracted from the LLM response
             tool_calls = yield ctx.call_activity(
                 self.get_tool_calls, input={"response": response}
             )
-
-            # Execute tool calls in parallel
-            if not ctx.is_replaying:
-                logger.info(f"Executing {len(tool_calls)} tool call(s)..")
-
-            parallel_tasks = [
-                ctx.call_activity(
-                    self.execute_tool,
-                    input={"instance_id": instance_id, "tool_call": tool_call},
-                )
-                for tool_call in tool_calls
-            ]
-            yield self.when_all(parallel_tasks)
-        else:
-            if not ctx.is_replaying:
-                logger.info("Agent generating response without tool execution..")
-
-            # No Tool Calls → Clear tools
-            self.tool_history.clear()
-
-        # Step 6: Determine if Workflow Should Continue
+            if tool_calls:
+                if not ctx.is_replaying:
+                    logger.debug(f"Executing {len(tool_calls)} tool call(s)..")
+                # Run the tool calls in parallel
+                parallel = [
+                    ctx.call_activity(self.run_tool, input={"tool_call": tc})
+                    for tc in tool_calls
```

Comment:
would there be a case a user wouldn't want tool calls run in parallel? Should this be optional?